sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/PartitionResolver.java

```diff
@@ -0,0 +1,231 @@
+import java.nio.ByteOrder;
+import java.nio.charset.StandardCharsets;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Allows events to be resolved to partitions using common patterns such as round-robin assignment and hashing of
+ * partitions keys.
+ */
+class PartitionResolver {
+    private static final ClientLogger LOGGER = new ClientLogger(PartitionResolver.class);
+    private static final int STARTING_INDEX = -1;
+
+    private final AtomicInteger partitionAssignmentIndex = new AtomicInteger(STARTING_INDEX);
+
+    /**
+     * Assigns a partition using a round-robin approach.
+     *
+     * @param partitions The set of available partitions.
+     *
+     * @return The zero-based index of the selected partition from the available set.
+     */
+    String assignRoundRobin(String[] partitions) {
+        Objects.requireNonNull(partitions, "'partitions' cannot be null.");
+
+        if (partitions.length == 0) {
+            throw LOGGER.logExceptionAsError(new IllegalArgumentException("'partitions' cannot be empty."));
+        }
+
+        final int currentIndex = partitionAssignmentIndex.accumulateAndGet(1,
+            (current, added) -> {
+                try {
+                    return Math.addExact(current, added);
+                } catch (ArithmeticException e) {
+                    LOGGER.info("Overflowed incrementing index. Rolling over.", e);
+
+                    return STARTING_INDEX + added;
+                }
+            });
+
+        return partitions[(currentIndex % partitions.length)];
+    }
+
+    /**
+     * Assigns a partition using a hash-based approach based on the provided {@code partitionKey}.
+     *
+     * @param partitionKey The partition key.
+     * @param partitions The set of available partitions.
+     *
+     * @return The zero-based index of the selected partition from the available set.
+     */
+    String assignForPartitionKey(String partitionKey, String[] partitions) {
+        final short hashValue = generateHashCode(partitionKey);
+        final int index = Math.abs(hashValue % partitions.length);
+
+        return partitions[index];
+    }
+
+    /**
+     * Generates a hashcode for the partition key using Jenkins' lookup3 algorithm.
+     *
+     * This implementation is a direct port of the Event Hubs service code; it is intended to match the gateway hashing
+     * algorithm as closely as possible and should not be adjusted without careful consideration.
+     *
+     * @param partitionKey The partition key.
+     *
+     * @return The generated hash code.
+     */
+    static short generateHashCode(String partitionKey) {
+        if (partitionKey == null) {
+            return 0;
+        }
+
+        final byte[] bytes = partitionKey.getBytes(StandardCharsets.UTF_8);
+
+        final Hashed hashed = computeHash(bytes, 0, 0);
+        final int i = hashed.getHash1() ^ hashed.getHash2();
+
+        return Integer.valueOf(i).shortValue();
```

Comment:
.NET cast uint to short, but here cast int to short. I afraid it will take the sign bit when it is a negative value, and the return value will be different? 
However, the the return value will be calculated by `Math.abs(hashValue % partitions.length)`, not sure eventually the index result is same.

Reply:
Yeah. I did some digging into this. Only matters if you're doing mathematical operations on the value. https://stackoverflow.com/a/9854205/4220757.